a10_octavia/controller/worker/controller_worker.py

```diff
@@ -661,27 +662,105 @@ def delete_member(self, member_id):
         finally:
             self._set_vthunder_available(member.project_id, True, ctx_flags, load_balancer)
 
+    def _is_batch_valid(self, old_member_ids, new_member_ids,
+                        updated_member_ids, member_collision_map):
+        valid = True
+        for mem_id, member_col in member_collision_map.items():
+            member, mem_cnt = member_col
+            mem_ip = member.ip_address
+            mem_port = member.protocol_port
+            if mem_cnt > 1:
+                if mem_id in old_member_ids:
+                    error_msg = ("Duplicate members with id {} and IP {} and port {} "
+                                 "found in member database.".format(mem_id, mem_ip, mem_port))
+                if mem_id in new_member_ids or mem_id in updated_member_ids:
+                    error_msg = ("Duplicate members with id {} and IP {} and port {} "
+                                 "found in batch update request.".format(mem_id, mem_ip, mem_port))
+                LOG.warning(error_msg)
+                valid = False
+        return valid
+
+    def _rollback_members(self, old_member_ids, new_member_ids, updated_member_ids):
+        set_o_ids = set(old_member_ids)
+        set_u_ids = set(updated_member_ids)
+
+        current_member_ids = set_o_ids.union(set_u_ids)
+
+        current_members = [self._member_repo.get(db_apis.get_session(), id=mid)
+                           for mid in current_member_ids]
+
+        for mem in current_members:
+            # Rollback status to prevent pending state lock
+            self._member_repo.update(db_apis.get_session(), mem.id,
+                                     provisioning_status=constants.ACTIVE)
+            LOG.info("Member with id {} and ip {} and port {} slated for "
+                     "batch update have been set to ACTIVE state.".format(
+                         mem.id, mem.ip_address, mem.protocol_port))
+
+        new_members = [self._member_repo.get(db_apis.get_session(), id=mid)
+                       for mid in new_member_ids]
+        for mem in new_members:
+            self._member_repo.delete(db_apis.get_session(), mem.id)
+            current_member_ids.add(mem.id)
+            LOG.info("Member with id {} and ip {} and port {} "
+                     "slated for creation under batch update "
+                     "has been deleted.".format(mem.id, mem.ip_address, mem.protocol_port))
+
+        modified_members = current_members + new_members
+        for mem in modified_members:
```

Comment:
Are all members in the same  pool and loadbalancer? or we need to do this for each members?

Reply:
I was concerned that we could have an edge case where the old members were under a different pool when initially writing the code. This concern was misplaced and after double checking the MembersController it isn't possible due to the filtering.

I've updated the code